contracts/src/BeefyClient.sol

```diff
@@ -391,48 +420,37 @@ contract BeefyClient {
-     *
-     * This function approximates f(x) defined below for x in [1, 21_846):
-     *
-     *  x <= 10: f(x) = x * (2/3)
-     *  x  > 10: f(x) = max(10, ceil(log2(3 * x))
-     *
-     * Research by W3F suggests that `ceil(log2(3 * x))` is a minimum number of signatures required to make an
-     * attack unfeasible. We put a further minimum bound of 10 on this value for extra security.
+     * @dev Calculates the number of signature samples required for `submitFinal`.
+     * @param validatorSetLen The validator set length.
+     * @param signatureUseCount The count of how many times a validators signature was previously used in a call to `submitInitial`.
      *
-     * If the session has less than 10 active validators it's definitely some sort of local testnet
-     * and we use different logic (minimum 2/3 + 1 validators must sign).
+     *  ceil(log2(validatorSetLen)) + 1 * 2 ceil(log2(signatureUseCount))
      *
-     * One assumption is that Polkadot/Kusama will never have more than roughly 20,000 active validators in a session.
-     * As of writing this comment, Polkadot has 300 validators and Kusama has around 1000 validators,
-     * so we are well within those limits.
-     *
-     * In any case, an order of magnitude increase in validator set sizes will likely require a re-architecture
-     * of Polkadot that would make this contract obsolete well before the assumption becomes a problem.
-     *
-     * Constants generated with the help of scripts/minsigs.py
+     * See https://hackmd.io/9OedC7icR5m-in_moUZ_WQ for full analysis.
      */
-    function minimumSignatureThreshold(uint256 validatorSetLen) internal pure returns (uint256) {
-        if (validatorSetLen <= 10) {
+    function signatureSamples(uint256 validatorSetLen, uint256 signatureUseCount) internal view returns (uint256) {
+        // There are less validators than the minimum signatures so validate 2/3 majority.
+        if (validatorSetLen <= minimumSignatureSamples) {
             return validatorSetLen - (validatorSetLen - 1) / 3;
-        } else if (validatorSetLen < 342) {
-            return 10;
-        } else if (validatorSetLen < 683) {
-            return 11;
-        } else if (validatorSetLen < 1366) {
-            return 12;
-        } else if (validatorSetLen < 2731) {
-            return 13;
-        } else if (validatorSetLen < 5462) {
-            return 14;
-        } else if (validatorSetLen < 10923) {
-            return 15;
-        } else if (validatorSetLen < 21846) {
-            return 16;
-        } else {
-            return 17;
         }
+
+        // Start with the minimum number of signatures.
+        uint256 samples = minimumSignatureSamples;
+
+        // We must substrate minimumSignatures from the number of validators or we might end up
+        // requiring more signatures than there are validators.
+        samples += Math.log2(validatorSetLen - minimumSignatureSamples, Math.Rounding.Ceil);
```

Comment:
```suggestion
        numRequiredSignatures += Math.log2(validatorSetLen - minNumRequiredSignatures, Math.Rounding.Ceil);
```

Reply:
Addressed in https://github.com/Snowfork/snowbridge/pull/956/commits/46250d8ff06a500a194a2c57acd59397e343af46